db/osqlshadtbl.c

```diff
@@ -602,20 +608,51 @@ int osql_fetch_shadblobs_by_genid(BtCursor *pCur, int *blobnum,
     key->seq = pCur->genid;
     key->id = *blobnum - 1;
 
-    rc = bdb_temp_table_find_exact(tbl->env->bdb_env, tbl->blb_cur, key,
-                                   sizeof(*key), bdberr);
+    /* We don't know the ODH-ness of the blob, so we search using
+     * bdb_temp_table_find(). */
+    rc = bdb_temp_table_find(tbl->env->bdb_env, tbl->blb_cur, key, sizeof(*key),
+                             NULL, bdberr);
     if (rc != IX_FND)
         free(key);
 
-    if (rc == IX_EMPTY || rc == IX_NOTFND) {
+    tmptblkey = bdb_temp_table_key(tbl->blb_cur);
+    if (rc == IX_EMPTY || rc == IX_NOTFND || key->seq != tmptblkey->seq ||
+        key->id != tmptblkey->id) {
         blobs->bloblens[0] = 0;
         blobs->bloboffs[0] = 0;
         blobs->blobptrs[0] = NULL;
         rc = 0;
     } else if (!rc) {
-        blobs->bloblens[0] = bdb_temp_table_datasize(tbl->blb_cur);
+        tmptblblb = bdb_temp_table_data(tbl->blb_cur);
+        tmptblblblen = bdb_temp_table_datasize(tbl->blb_cur);
+
+        if (tmptblkey->odh) {
+            rc = bdb_unpack_heap(pCur->db->handle, tmptblblb, tmptblblblen,
+                                 (void **)blobs->blobptrs, blobs->bloblens,
+                                 &freeptr);
+            if (rc != 0) {
+                free(freeptr);
```

Comment:
Do we double free here? Looks like on errors, `bdb_unpack_heap` also frees `freeptr`

Reply:
Yes we do - thanks for catching this!